pkg/castemplate/v1alpha1/kubernetes_test.go

```diff
@@ -0,0 +1,220 @@
+			}
+
+			fc.withDefaults()
+			if mock.expectGet && fc.get == nil {
+				t.Fatalf(`test %s failed: expected non-nil fc.get
+but got %v`, name, fc.get)
+			}
+			if mock.expectGetClientset && fc.getClientset == nil {
+				t.Fatalf(`test %s failed: expected non-nil fc.getClientset
+but got %v`, name, fc.getClientset)
+			}
+		})
+	}
+}
+func TestWithClientset(t *testing.T) {
+	tests := map[string]struct {
+		clientset    *clientset.Clientset
+		isKubeClient bool
+	}{
+		"Clientset is empty":     {nil, false},
+		"Clientset is not empty": {&clientset.Clientset{}, true},
+	}
+
+	for name, mock := range tests {
+		t.Run(name, func(t *testing.T) {
+			kc := KubeClient(WithClientset(mock.clientset))
+			if mock.isKubeClient && kc.clientset == nil {
+				t.Fatalf(`test %s failed, expected non-nil fake.clientset
+but got %v`, name, kc.clientset)
+			}
+			if !mock.isKubeClient && kc.clientset != nil {
+				t.Fatalf(`test %s failed, expected nil fake.clientset
+but got %v`, name, kc.clientset)
+			}
+		})
+	}
+}
+func TestKubeClientWithClientset(t *testing.T) {
+	tests := map[string]struct {
+		opts            []KubeclientBuildOption
+		expectClientSet bool
+	}{
+		"When non-nil clientset is passed": {
+			[]KubeclientBuildOption{fakeSetClientsetOk}, true},
+		"When two options with a non-nil clientset are passed": {
+			[]KubeclientBuildOption{fakeSetClientsetOk, fakeClientSet}, true},
+		"When three options with a non-nil clientset are passed": {
+			[]KubeclientBuildOption{fakeSetClientsetOk, fakeClientSet, fakeClientSet}, true},
+		"When nil clientset is passed": {
+			[]KubeclientBuildOption{fakeSetClientsetNil}, false},
+		"When two options with a nil clientset are passed": {
+			[]KubeclientBuildOption{fakeSetClientsetNil, fakeClientSet}, false},
+		"When three options with a nil clientset are passed": {
+			[]KubeclientBuildOption{fakeSetClientsetNil, fakeClientSet, fakeClientSet}, false},
+	}
+
+	for name, mock := range tests {
+		t.Run(name, func(t *testing.T) {
+			c := KubeClient(mock.opts...)
+			if !mock.expectClientSet && c.clientset != nil {
+				t.Fatalf(`test %s failed, expected nil c.clientset
+but got %v`, name, c.clientset)
+			}
+			if mock.expectClientSet && c.clientset == nil {
+				t.Fatalf(`test %s failed expected non-nil c.clientset
+but got %v`, name, c.clientset)
+			}
+		})
+	}
+}
+
+func TestGetClientOrCached(t *testing.T) {
+
+	tests := map[string]struct {
+		clientset    *clientset.Clientset
+		getClientset getClientsetFunc
+		get          getFunc
+		expectErr    bool
+	}{
+		// Positive tests
+		"When clientset is nil":     {nil, fakeGetClientsetNil, fakeGetOk, false},
+		"When clientset is not nil": {&clientset.Clientset{}, fakeGetClientsetNil, fakeGetOk, false},
+		// Negative tests
+		"When getting clientset throws error": {nil, fakeGetClientsetErr, fakeGetOk, true},
+	}
+
+	for name, mock := range tests {
+		t.Run(name, func(t *testing.T) {
+			kc := KubeClient(WithClientset(mock.clientset),
```

Comment:
IMO this can be made more readable:
```go
kc := &Kubeclient{
  clientset: mock.clientset,
  getClientset: mock.getClientset,
  get: mock.get,
}
```

Reply:
done